cpp/distributed/dist-mnist.cpp

```diff
@@ -0,0 +1,192 @@
+struct Model : torch::nn::Module {
+  Model()
+      : conv1(torch::nn::Conv2dOptions(1, 10, 5)),
+        conv2(torch::nn::Conv2dOptions(10, 20, 5)),
+        fc1(320, 50),
+        fc2(50, 10) {
+    register_module("conv1", conv1);
+    register_module("conv2", conv2);
+    register_module("conv2_drop", conv2_drop);
+    register_module("fc1", fc1);
+    register_module("fc2", fc2);
+  }
+
+  torch::Tensor forward(torch::Tensor x) {
+    x = torch::relu(torch::max_pool2d(conv1->forward(x), 2));
+    x = torch::relu(
+        torch::max_pool2d(conv2_drop->forward(conv2->forward(x)), 2));
+    x = x.view({-1, 320});
+    x = torch::relu(fc1->forward(x));
+    x = torch::dropout(x, 0.5, is_training());
+    x = fc2->forward(x);
+    return torch::log_softmax(x, 1);
+  }
+
+  torch::nn::Conv2d conv1;
+  torch::nn::Conv2d conv2;
+  torch::nn::Dropout2d conv2_drop;
+  torch::nn::Linear fc1;
+  torch::nn::Linear fc2;
+};
+
+int main(int argc, char* argv[]) {
+  // MPI variables
+  int rank, numranks;
+  MPI_Init(&argc, &argv);
+  MPI_Comm_size(MPI_COMM_WORLD, &numranks);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  // end MPI variables
+
+  // Timer variables
+  auto tstart = 0.0;
+  auto tend = 0.0;
+
+  // TRAINING
+  // Read train dataset
+  const char* kDataRoot = "../data";
+  auto train_dataset =
+      torch::data::datasets::MNIST(kDataRoot)
+          .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
+          .map(torch::data::transforms::Stack<>());
+
+  // Distributed Random Sampler
+  auto data_sampler = torch::data::samplers::DistributedRandomSampler(
+      train_dataset.size().value(), numranks, rank, false);
+
+  auto num_train_samples_per_proc = train_dataset.size().value() / numranks;
+
+  // Generate dataloader
+  auto total_batch_size = 64;
+  auto batch_size_per_proc =
+      total_batch_size / numranks; // effective batch size in each processor
+  auto data_loader = torch::data::make_data_loader(
+      std::move(train_dataset), data_sampler, batch_size_per_proc);
+
+  // setting manual seed
+  torch::manual_seed(0);
+
+  auto model = std::make_shared<Model>();
+
+  auto learning_rate = 1e-2;
+
+  torch::optim::SGD optimizer(model->parameters(), learning_rate);
+
+  // Number of epochs
+  size_t num_epochs = 10;
+
+  // start timer
+  tstart = MPI_Wtime();
+
+  for (size_t epoch = 1; epoch <= num_epochs; ++epoch) {
+    size_t num_correct = 0;
+
+    for (auto& batch : *data_loader) {
+      auto ip = batch.data;
+      auto op = batch.target.squeeze();
+
+      // convert to required formats
+      ip = ip.to(torch::kF32);
+      op = op.to(torch::kLong);
+
+      // Reset gradients
+      model->zero_grad();
+
+      // Execute forward pass
+      auto prediction = model->forward(ip);
+
+      auto loss = torch::nll_loss(torch::log_softmax(prediction, 1), op);
+
+      // Backpropagation
+      loss.backward();
+
+      // Averaging the gradients of the parameters in all the processors
```

Comment:
I'm wondering whether we should add a comment here indicating that while this approach will allow synchronous data parallel training, it will lag behind DDP in performance since it synchronizes parameters after the backward pass whereas DDP overlaps synchronizing parameters and computing gradients in the backwards pass.

Reply:
@osalpekar I can add that comment. Also I can call MPI_Iallreduce before the backward pass which is a non-blocking call, then call MPI_Wait for that call after the backward pass. This will ensure that backward pass and synchronizing parameters are overlapped. What do you say?

Reply:
I don't think that will have the same effect here. `loss.backward` is what actually computes the gradients so I don't think we can launch the synchronization collective before we've even started computing gradients.

DDP has a notion of buckets. We've instrumented the backward pass in DDP such that, once a bucket has been filled with gradients, we launch the allreduce. While the allreduce call is executing, we continue computing the next gradients to fill the next bucket, and so on. Trying to re-create this behavior efficiently is probably beyond the scope of this example so we can just stick to a blocking allreduce after the backward pass, but just in case somebody tries to measure the performance, we should include the comment so that they know this approach will be slower than DDP.

Reply:
I see! Added the comment :)